base/compiler/abstractinterpretation.jl

```diff
@@ -1033,9 +1033,35 @@ function abstract_call(interp::AbstractInterpreter, fargs::Union{Nothing,Vector{
             add_remark!(interp, sv, "Could not identify method table for call")
             return CallMeta(Any, false)
         end
-        return abstract_call_gf_by_type(interp, nothing, argtypes, argtypes_to_type(argtypes), sv, max_methods)
+        callinfo = abstract_call_gf_by_type(interp, nothing, argtypes, argtypes_to_type(argtypes), sv, max_methods)
+        return callinfo_from_interprocedural(callinfo, fargs)
+    end
+    callinfo = abstract_call_known(interp, f, fargs, argtypes, sv, max_methods)
+    return callinfo_from_interprocedural(callinfo, fargs)
+end
+
+function callinfo_from_interprocedural(callinfo::CallMeta, ea::Union{Nothing,Vector{Any}})
+    if ea === nothing
+        return callinfo
+    end
+
+    rt = callinfo.rt
+    if isa(rt, InterConditional)
+        # try to convert interprocedural conditional constraint from callee into the constraint
+        # on slots of the current frame
+        i = rt.slot
+        if checkbounds(Bool, ea, i)
+            # `InterConditional` is supposed to be passed here only after "valid" `abstract_call`,
+            # and thus `i` should always be in bounds of `ea`, but just for safety
+            e = @inbounds ea[i]
+            if isa(e, Slot)
+                return CallMeta(Conditional(e, rt.vtype, rt.elsetype), callinfo.info)
+            end
+        end
+        return CallMeta(widenconditional(rt), callinfo.info)
+    else
```

Comment:
Is it okay to have `@assert`s for this kind of (somewhat performance-critical) code ? 
I have a naive assumption that `throw` calls lead to worse performance.  Maybe only `try/catch` matters in performance aspect, but `throw` doesn't ?

Reply:
try/catch is slow, but the existence of unreached code here is (nearly) free

Reply:
okay, let's assert we don't mix-up `Conditional` in inter-procedural context.

Well, `@assert` shouldn't be here, since `abstract_call_builtin` can return `Conditional`s. 
I'd do this assertion in `abstract_call_gf_by_type`): 25c5f151e791b2bd8a0a6c3f11c6e2171504a5cc